In [1]:
import pandas as pd
pd.options.display.float_format = '{:.8e}'.format
from population import generate_init
from character import CharacterType
#from src.stats import random_stats

N = 200

maxStatsValue = 150
type = CharacterType.ARCHER

#equipment_data = random_stats(N, maxStatsValue)
poblacion_inicial = generate_init(N,maxStatsValue,type)

#poblacion_inicial


In [2]:
#Añado las columnas de desempeño relativo y acumulado
from selection import add_relative_accumulate
poblacion_rel_acu = add_relative_accumulate(poblacion_inicial)
poblacion_rel_acu

sum: 52519.94526990462


strength  agility  expertise  resistance  life         height  \
30         59       10         70           4     7 1.90000000e+00   
184        64       16         59          10     1 1.40000000e+00   
27         67        6         51          13    13 1.80000000e+00   
157        57       13         55           8    17 1.90000000e+00   
105        46        2         71          16    15 1.80000000e+00   
..        ...      ...        ...         ...   ...            ...   
4           4      109          2           9    26 1.40000000e+00   
89          4       11          7          41    87 1.50000000e+00   
167         2       79         10          55     4 1.70000000e+00   
13          5       20          2         101    22 1.40000000e+00   
31          4       18          1          20   107 1.40000000e+00   

                     type    performance  performance_relative  \
30   CharacterType.ARCHER 1.10370534e+03        2.10149751e-02   
184  CharacterType.ARCHER 1.00678852e+03        1.91696414e-02   
27   CharacterType.ARCHER 9.18530409e+02        1.74891730e-02   
157  CharacterType.ARCHER 9.04922869e+02        1.72300802e-02   
105  CharacterType.ARCHER 8.66007004e+02        1.64891071e-02   
..                    ...            ...                   ...   
4    CharacterType.ARCHER 2.01345465e+01        3.83369525e-04   
89   CharacterType.ARCHER 1.73183697e+01        3.29748433e-04   
167  CharacterType.ARCHER 1.37513995e+01        2.61831946e-04   
13   CharacterType.ARCHER 1.05135331e+01        2.00181722e-04   
31   CharacterType.ARCHER 7.25342158e+00        1.38107943e-04   

     performance_accumulated  
30            2.10149751e-02  
184           4.01846165e-02  
27            5.76737895e-02  
157           7.49038697e-02  
105           9.13929768e-02  
..                       ...  
4             9.99070130e-01  
89            9.99399878e-01  
167           9.99661710e-01  
13            9.99861892e-01  
31            1.00000000e+00  

[200 rows x 10 columns]

#### Seleccion de k indiduos
A*(método1) + (1-A)*(método2)

In [3]:
import numpy as np
#Dado un K
#Si quiero un m1_perc de la selección con metodo1
K = 120
m1_perc = 0.3
N1 = int(np.round(K * m1_perc))
N2 = K - N1
#N1

In [4]:
from selection import elitist_selection
nueva_poblacion_e = elitist_selection(poblacion_rel_acu,N1)
#nueva_poblacion_e

In [5]:
from selection import roulette_wheel_selection
nueva_poblacion_rul = roulette_wheel_selection(poblacion_rel_acu, N2)
#nueva_poblacion_rul

In [6]:
import pandas as pd
selection_total = pd.concat([nueva_poblacion_e,nueva_poblacion_rul])
selection_total

strength agility expertise resistance life         height  \
0        59      10        70          4    7 1.90000000e+00   
1        64      16        59         10    1 1.40000000e+00   
2        67       6        51         13   13 1.80000000e+00   
3        57      13        55          8   17 1.90000000e+00   
4        46       2        71         16   15 1.80000000e+00   
..      ...     ...       ...        ...  ...            ...   
79       49       9        33         57    2 2.00000000e+00   
80       30      32        46         38    4 1.60000000e+00   
81       49       9        33         57    2 2.00000000e+00   
82       47      47        17         24   15 1.40000000e+00   
83        7      35        51         11   46 1.90000000e+00   

                    type    performance performance_relative  \
0   CharacterType.ARCHER 1.10370534e+03       2.10149751e-02   
1   CharacterType.ARCHER 1.00678852e+03       1.91696414e-02   
2   CharacterType.ARCHER 9.18530409e+02       1.74891730e-02   
3   CharacterType.ARCHER 9.04922869e+02       1.72300802e-02   
4   CharacterType.ARCHER 8.66007004e+02       1.64891071e-02   
..                   ...            ...                  ...   
79  CharacterType.ARCHER 4.91255423e+02       9.35369259e-03   
80  CharacterType.ARCHER 4.43746636e+02       8.44910698e-03   
81  CharacterType.ARCHER 4.91255423e+02       9.35369259e-03   
82  CharacterType.ARCHER 3.40759668e+02       6.48819542e-03   
83  CharacterType.ARCHER 1.33152028e+02       2.53526594e-03   

   performance_accumulated  
0           2.10149751e-02  
1           4.01846165e-02  
2           5.76737895e-02  
3           7.49038697e-02  
4           9.13929768e-02  
..                     ...  
79          3.92694624e-01  
80          4.27784511e-01  
81          3.92694624e-01  
82          5.78408861e-01  
83          9.03180951e-01  

[120 rows x 10 columns]

In [7]:
from genes import  encode_genes

cromosomas = encode_genes(selection_total)
#cromosomas

In [8]:
from crossover import single_point_crossover, valid_chromosome

new_individuals = single_point_crossover(cromosomas)
cromosomas_ok1 = valid_chromosome(new_individuals)
while (len(cromosomas_ok1)<1):
    new_individuals = single_point_crossover(cromosomas)
    cromosomas_ok1 = valid_chromosome(new_individuals)
#new_individuals
cromosomas_ok1

strength   agility expertise resistance      life height
28   00010001  00000101  00110101   00110110  00010101    011
29   01100100  00000010  00010100   00011000  00000100    000
72   00010100  00001010  00111110   00110111  00000011    011
73   00101100  00111110  00010011   00001101  00001100    000
86   01010001  00000101  00010101   00010110  00010101    101
87   00010110  00101010  00111101   00001100  00001101    010
98   00110110  00000100  00111100   00001001  00010111    100
99   00010111  00000100  00111010   00100011  00011110    000
106  01001000  00011011  00001010   00011010  00001111    100
107  01010001  00000011  00011110   00001100  00011000    110

In [9]:
#from genes import decode_genes
#cromosomas_decoded = decode_genes(new_individuals)
#cromosomas_decoded

- Cantidad de hijos validos: 4
-     strength  agility expertise	resistance	life  height  type	                performance	 performance_relative   performance_accumulated
- 5	  2	        30	    3	        6	        109	  2.0	  CharacterType.ARCHER	6.277699	 0.001625	            1.0
- 6	  50	    18	    3	        18	        61	  1.7	  CharacterType.ARCHER	119.533975	 0.030933	            0.881345
- 9	  4	        39	    92	        4	        11	  1.4	  CharacterType.ARCHER	106.010218	 0.027433	            0.908778
- 10  4	        39	    92	        4	        11	  1.4	  CharacterType.ARCHER	106.010218	 0.027433	            0.908778

In [10]:
from crossover import two_point_crossover, valid_chromosome

new_individuals2 = two_point_crossover(cromosomas)
cromosomas_ok2 = valid_chromosome(new_individuals2)

while (len(cromosomas_ok2)<1):
    new_individuals2 = single_point_crossover(cromosomas)
    cromosomas_ok2 = valid_chromosome(new_individuals2)
#new_individuals2
cromosomas_ok2

strength   agility expertise resistance      life height
34   00010111  00101000  00111101   00001110  00001100    001
35   00011000  00000110  00111101   00101100  00001111    100
38   00110000  00011111  00110100   00001100  00000111    001
39   00101100  00100100  00110011   00000111  00001100    000
48   00110100  00000010  00100100   00111000  00000100    010
49   00100000  01000000  00010111   00011011  00000100    000
100  00011100  00010000  01100001   00000100  00000101    010
101  01001000  00001000  00001110   00001001  00101111    111

In [11]:
#from genes import decode_genes
#cromosomas_decoded2 = decode_genes(new_individuals2)
#cromosomas_decoded2

In [12]:
#from genes import decode_genes
#cromosomas_decoded2 = decode_genes(cromosomas_ok)
#cromosomas_decoded2

In [18]:
from mutation import one_gene_mutation
from crossover import valid_chromosome

mutation_rate = 0.1
if( len(cromosomas_ok1) > len(cromosomas_ok2) ):

    print("single_point_crossover won")
    result = one_gene_mutation(cromosomas_ok1, mutation_rate)
    cromosomas_okm = valid_chromosome(result)

    while (len(cromosomas_okm)<1):
        result = one_gene_mutation(cromosomas_ok1, mutation_rate)
        cromosomas_okm = valid_chromosome(result)
else:     
    print("two_point_crossover won")   
    result = one_gene_mutation(cromosomas_ok2, mutation_rate)
    cromosomas_okm = valid_chromosome(result)

    while (len(cromosomas_okm)<1):
        result = one_gene_mutation(cromosomas_ok2, mutation_rate)
        cromosomas_okm = valid_chromosome(result)


cromosomas_okm



single_point_crossover won


strength   agility expertise resistance      life height
29  01100100  00000010  00010100   00011000  00000100    001
72  00010100  00001010  00111110   00110111  00000011    011
73  00101100  00111110  00010011   00001101  00001100    000
86  01010001  00000101  00010101   00010110  00010101    101
98  00110110  00000100  00111100   00001001  00010111    100
99  00010111  00000100  00111010   00100011  00011110    000

In [19]:
from genes import decode_genes
cromosomas_decoded2 = decode_genes(cromosomas_okm)
cromosomas_decoded2

strength  agility  expertise  resistance  life         height
1       100        2         20          24     4 1.40000000e+00
2        20       10         62          55     3 1.60000000e+00
3        44       62         19          13    12 1.30000000e+00
4        81        5         21          22    21 1.80000000e+00
5        54        4         60           9    23 1.70000000e+00
6        23        4         58          35    30 1.30000000e+00

In [20]:
from population import eval_performace

new_generation = eval_performace(cromosomas_decoded2,type)
new_generation

strength  agility  expertise  resistance  life         height  \
5        54        4         60           9    23 1.70000000e+00   
1       100        2         20          24     4 1.40000000e+00   
4        81        5         21          22    21 1.80000000e+00   
6        23        4         58          35    30 1.30000000e+00   
2        20       10         62          55     3 1.60000000e+00   
3        44       62         19          13    12 1.30000000e+00   

                   type    performance  
5  CharacterType.ARCHER 8.56718352e+02  
1  CharacterType.ARCHER 4.90122644e+02  
4  CharacterType.ARCHER 4.72664284e+02  
6  CharacterType.ARCHER 3.64841039e+02  
2  CharacterType.ARCHER 3.54614831e+02  
3  CharacterType.ARCHER 3.39400618e+02

In [73]:
import numpy as np
from selection import boltzmann_selection

# Example usage:

max_generation = 18
generation = 1
k = 4                                  # Number of desired selected individuals.
T_0 = 20                                 # Initial temperature.
T_offset = 0.5                          # Offset value for temperature exponential function.
reference_generation = max_generation   # Reference generation where 0 < (T_0 - T_offset) = 0.015 = deltaT is desired.
deltaT = 0.015                          # 0 < (T_0 - T_offset) = deltaT at reference generation.
m = - (np.log(deltaT) - np.log(T_0 - T_offset)) / reference_generation

new_boltzmann_selection = boltzmann_selection(new_generation,k,generation,T_0,T_offset,m)

new_boltzmann_selection


sum: 0.9999999999999999
   strength  agility  expertise  resistance  life         height  \
5        54        4         60           9    23 1.70000000e+00   
1       100        2         20          24     4 1.40000000e+00   
4        81        5         21          22    21 1.80000000e+00   
6        23        4         58          35    30 1.30000000e+00   
2        20       10         62          55     3 1.60000000e+00   
3        44       62         19          13    12 1.30000000e+00   

                   type    performance  performance_relative  \
5  CharacterType.ARCHER 1.00000000e+00        1.00000000e+00   
1  CharacterType.ARCHER 1.93767478e-12        1.93767478e-12   
4  CharacterType.ARCHER 5.36397744e-13        5.36397744e-13   
6  CharacterType.ARCHER 1.92547688e-16        1.92547688e-16   
2  CharacterType.ARCHER 9.07426300e-17        9.07426300e-17   
3  CharacterType.ARCHER 2.96290571e-17        2.96290571e-17   

   performance_accumulated  
5           1.0000000

strength  agility  expertise  resistance  life         height  \
0        54        4         60           9    23 1.70000000e+00   
1        54        4         60           9    23 1.70000000e+00   
2        54        4         60           9    23 1.70000000e+00   
3        54        4         60           9    23 1.70000000e+00   

                   type    performance  performance_relative  \
0  CharacterType.ARCHER 1.00000000e+00        1.00000000e+00   
1  CharacterType.ARCHER 1.00000000e+00        1.00000000e+00   
2  CharacterType.ARCHER 1.00000000e+00        1.00000000e+00   
3  CharacterType.ARCHER 1.00000000e+00        1.00000000e+00   

   performance_accumulated  
0           1.00000000e+00  
1           1.00000000e+00  
2           1.00000000e+00  
3           1.00000000e+00